### PR TITLE
fix: demote disconnect-caused stateless session crash logs

### PR DESCRIPTION
**Pull request by swissmo**
## What does this PR do?

Every ha-mcp HTTP launcher (CLI, the Home Assistant app, and the in-process embedded server) runs Streamable HTTP in stateless mode. When a tool call outlives the client's patience — a slow backend call, a client-side timeout, or an unresponsive client for any other reason — the `mcp` SDK's `StreamableHTTPSessionManager` finishes serving the HTTP response and tears the transport down while the app is still working; the eventual attempt to deliver the response then writes into an already-closed memory stream, raising `anyio.ClosedResourceError` wrapped in a doubly-nested `ExceptionGroup`. The SDK already catches this — it's not an unhandled crash, and there's no user-visible failure beyond the response the client already gave up on — but it logs a full ERROR-level traceback under `mcp.server.streamable_http_manager`, which reads as a server bug when it's actually an expected race in a stateless HTTP protocol.

This PR adds `SessionDisconnectLogFilter` (new `src/ha_mcp/log_filters.py`), following the existing `StatelessSessionLogFilter`/`ToolValidationLogFilter` pattern already used for similar SDK/fastmcp log noise: it recursively unwraps (possibly nested) `ExceptionGroup`s and demotes the log entry to a one-line WARNING only when every leaf is a `ClosedResourceError`; a mixed group (a real bug alongside the disconnect race) stays at ERROR with its traceback intact. The three filter classes and a shared `install_sdk_log_filters()` installer were extracted into this new, side-effect-free module (mirroring `browser_landing.py`'s reasoning) so the in-process embedded server can safely call it too — previously only the CLI's `_setup_logging()` wired any of these filters, so the embedded server got none of them, not just the new one.

**Live-validated end-to-end (2026-08-25)** against the real production scenario that originally reported this: the exact `ExceptionGroup(ExceptionGroup(ClosedResourceError()))` shape is correctly recognized and demoted to `Stateless session crashed: client disconnected before response delivery`, with no traceback.

## Type of change
- [x] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 📚 Documentation
- [ ] 🔧 Maintenance/refactor
- [ ] 🧪 Tests only
- [ ] 💥 Breaking change

## Testing
- [x] I have tested these changes with a LLM agent
- [x] All automated tests pass (`uv run pytest`)
- [x] Code follows style guidelines (`uv run ruff check`)

## Future improvements

The underlying reason a tool call can take long enough to trigger this disconnect race in the first place (a ~60s hang/timeout reported against a `search_media` call in a real voice-assistant scenario) is being investigated separately — that's a latency/root-cause question in a different part of the stack, not a log-formatting one, and is out of scope for this PR.

## Checklist
- [x] I have updated documentation if needed

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of expected client disconnects during streaming sessions, including nested teardown errors.
  * Downgraded routine disconnect and tool-validation errors to warnings without misleading tracebacks.
  * Preserved full error details for unexpected failures.
  * Improved compatibility with older installations and externally managed setups.
  * Prevented duplicate logging filters during repeated startup or configuration.

* **Tests**
  * Added coverage for disconnect logging, nested exception handling, startup compatibility, package validation, and logging configuration cleanup.

### CODE REVIEW

**Comment by ghhamcp** [comment]
@codex review — apply the review criteria in .gemini/styleguide.md in addition to AGENTS.md guidance

File: src/ha_mcp/__main__.py [hunk 1]
```diff
@@ -446,6 +447,43 @@ def filter(self, record: logging.LogRecord) -> bool:
...
+    known-benign failures. Any other exception on this logger -- an actual
+    crash -- is left untouched.
+    """
+
+    def filter(self, record: logging.LogRecord) -> bool:
+        if record.name != "mcp.server.streamable_http_manager" or not record.exc_info:
+            return True
+
+        err = record.exc_info[1]
+        if not isinstance(err, anyio.ClosedResourceError):
```
[the remark this answers is not in the document]
Comment (swissmo) [reply src/ha_mcp/__main__.py]: Confirmed and fixed. Verified against the real code path: mcp.server.lowlevel.server.Server.run()
dispatches _handle_message via anyio.create_task_group().start_soon(...), and fastmcp's own
run() (fastmcp/server/low_level.py) wraps that in a second task group — so a ClosedResourceError
from _send_response arrives here as a doubly-nested ExceptionGroup, not a bare exception (matches
what showed up in a live production log during testing).

Added `_is_only_closed_resource_errors()` in `src/ha_mcp/log_filters.py`, which recursively
unwraps (possibly nested) ExceptionGroups and only demotes when every leaf is a
ClosedResourceError. A mixed group (a real bug alongside the disconnect race) stays at ERROR
with its traceback intact.

Regression coverage now reproduces the actual task-group boundary with a live
`anyio.create_task_group()` child raising the exception, rather than a hand-built `exc_info`
(`test_demotes_real_task_group_exception_group`, `test_leaves_mixed_exception_group_at_error`
in `tests/src/unit/test_session_disconnect_log_filter.py`), plus direct unit coverage of the
recursive helper itself (nested groups, mixed groups, empty-group edge case).

File: src/ha_mcp/__main__.py [hunk 2]
```diff
@@ -531,6 +569,9 @@ def _setup_logging(log_level_str: str, force: bool = True) -> None:
     logging.getLogger("mcp.server.streamable_http").addFilter(
         StatelessSessionLogFilter()
     )
+    logging.getLogger("mcp.server.streamable_http_manager").addFilter(
+        SessionDisconnectLogFilter()
+    )
```
[the remark this answers is not in the document]
Comment (swissmo) [reply src/ha_mcp/__main__.py]: Confirmed and fixed — and broader than described: the embedded server installed none of the
three SDK/fastmcp log-noise filters, not just the new one.

Extracted all three filters (StatelessSessionLogFilter, ToolValidationLogFilter,
SessionDisconnectLogFilter) plus a shared `install_sdk_log_filters()` into a new
`src/ha_mcp/log_filters.py` module — side-effect-free, so it's safe to import from the
in-process embedded server (importing `ha_mcp.__main__` directly would run process-global
side effects like truststore SSL patching, which must never happen in-process — same
reasoning `ha_mcp/browser_landing.py`'s docstring already documents for the landing-page
helper).

- `ha_mcp.__main__._setup_logging()` now calls `install_sdk_log_filters()`.
- `homeassistant-addon/start.py` now calls it too (previously only wired
  StatelessSessionLogFilter manually).
- `custom_components/ha_mcp_tools/embedded_server.py::_serve()` now calls it, guarded by the
  same ImportError fallback pattern as the existing `register_browser_landing` import just
  above it (the component and server package version independently, so an older installed
  ha-mcp may predate `ha_mcp.log_filters` entirely).

New hermetic tests in `tests/src/unit/test_embedded_server.py`
(`TestServeLogFilters::test_serve_installs_log_filters` and
`test_serve_tolerates_missing_log_filters_module`) cover both the install and the
backward-compat fallback, mirroring the existing browser-landing test pair exactly.

File: src/ha_mcp/log_filters.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    ``install_sdk_log_filters()`` can run more than once per process: the
+    in-process embedded server calls it on every ``_serve()`` reload without
+    a process restart, and process-wide ``logging`` state (including each
+    named logger's filter list) persists across those reloads. Without this
+    guard, filters would accumulate one more redundant instance per reload.
+    """
+    logger = logging.getLogger(logger_name)
+    if any(isinstance(f, filter_cls) for f in logger.filters):
+        return
+    logger.addFilter(filter_cls())
```
[the remark this answers is not in the document]
Comment (swissmo) [reply src/ha_mcp/log_filters.py:193]: CodeRabbit inline finding (fixed): _add_filter_once's dedup check was comparing by isinstance(), which only works within a single loaded generation of the module. Since the embedded server purges and re-imports ha_mcp.* fresh on every reinstall, each reload gets a brand-new SessionDisconnectLogFilter class object — isinstance() against a stale-generation instance never matches, so filters would've kept accumulating one more stale copy per reload forever. Fixed to match by (module, qualname) instead (stable across reloads) and actively replace stale instances rather than just skip.

Worth flagging: my first attempt at the regression test used importlib.reload() to simulate this, which actually broke two other test files — reloading a module mutates shared global state, so files that had already imported the old class reference at collection time started failing isinstance checks against the newly-installed instances. Caught it by running the full affected test set with --maxfail=0 before committing, and rewrote the test to build a synthetic stale-generation class instead (no reload, no shared-state risk).

CodeRabbit nitpick (addressed, scoped down): wanted full launcher-level e2e coverage for all three HTTP launchers. Docker isn't available in this environment, so I can't run the existing Docker-based add-on e2e suite (tests/addon/test_addon_startup.py) locally — but it already exists and will run automatically via this PR's CI (haos-e2e-inaddon-tests.yml etc.), so I didn't bolt new heavy hermetic mocking onto start.py's 900-line main() just to duplicate that coverage locally.

File: src/ha_mcp/log_filters.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        # shape it saw and how it classified it.
+        _diag_logger.warning(
+            "SessionDisconnectLogFilter saw exc_info on %s: type=%r module=%r "
+            "repr=%.200r classified=%s",
+            record.name,
+            type(err).__name__,
+            type(err).__module__,
+            err,
+            classified,
+        )
```
[the remark this answers is not in the document]
Comment (swissmo) [reply src/ha_mcp/log_filters.py]: The temporary diagnostics (`_diag_logger.warning` in `SessionDisconnectLogFilter.filter()` and in `install_sdk_log_filters()`) were added specifically to track down why the fix wasn't taking effect in a live deployment, and were removed in commit 6f586a82 once that was root-caused and the fix was confirmed working end-to-end against the real production scenario. Current code has no diagnostic warnings left — just the normal per-record demotion and the plain `_add_filter_once` installer logic.

File: custom_components/ha_mcp_tools/embedded_server.py [hunk 5]
```diff
@@ -244,6 +244,56 @@ def _worker_startup_failure(exc: BaseException) -> EmbeddedServerError:
...
+            err,
+        )
+        return
+    except ImportError as err:
+        _LOGGER.warning(
+            "Could not install MCP SDK log-noise filters; continuing without them: %s",
+            err,
+        )
+        return
+    _LOGGER.warning("_install_log_filters_if_available: import succeeded, installing")
```
[the remark this answers is not in the document]
Comment (swissmo) [reply custom_components/ha_mcp_tools/embedded_server.py]: Resolved in 6f586a82 — the entry/ModuleNotFoundError/successful-import diagnostic warnings in `_install_log_filters_if_available()` were temporary instrumentation for the same live-debugging effort and have been removed. Current code is back to the original shape: silent on the genuine "older ha-mcp, module doesn't exist" case, a WARNING only for a real unexpected failure (a different `ModuleNotFoundError`, a plain `ImportError`, or an exception from `install_sdk_log_filters()` itself) — no new warning-level startup logs.

**Comment by ghhamcp** [comment]
@codex review — apply the review criteria in .gemini/styleguide.md in addition to AGENTS.md guidance

**Review by Patch76 — COMMENTED** [review]
The recursive leaf test is the right shape here, and the mixed-group case staying at ERROR is what makes it safe to ship: a demotion rule that swallowed a real bug travelling alongside the disconnect would be worse than the noise it removes. Pulling all three filters into a side-effect-free module rather than only the new one is also the right call, since the embedded server was missing all of them.

Two things came out of checking the classifier against the pinned SDK.

**[Concern 1]: at `mcp==1.28.1` I could not find the send that produces this record.**

The response delivery the PR body describes is handled inside the SDK at the pinned version: `Server._handle_request` wraps `await message.respond(response)` in `except (anyio.BrokenResourceError, anyio.ClosedResourceError)` and drops it at DEBUG (`mcp/server/lowlevel/server.py:805-815`). At `mcp==1.24.0` that line is bare; across the releases I have on hand the guard is absent at 1.24.0 and present from 1.27.2 onwards, so it landed somewhere in between.

Walking the five `_write_stream.send` sites in `mcp/shared/session.py` by caller rather than by site, every path a slow tool call can take is swallowed before it could reach `logger.exception("Stateless session crashed")`. `send_request` (`mcp/shared/session.py:281`) and a notification raised from inside a tool never escape the handler, with `mcp/server/lowlevel/server.py:791-794` as the backstop that turns whatever emerges into an error response (`raise_exceptions` defaults to False, and neither manager call site passes it). `_send_response` reached through `respond()` (`:134`) sits behind the guard above; reached through `cancel()` (`:148`), and the validation-error response at `:395`, both sit inside `_receive_loop`, which swallows everything twice over (`:428`, `:436`, `:441`). Also worth noting, ha-mcp's own `ctx.report_progress` / `ctx.info` calls are already wrapped in `safe_progress` / `safe_info` (`src/ha_mcp/tools/helpers.py:586-608`). The one caller I found still unguarded is `_handle_message`'s `case Exception()` branch, which calls `send_log_message` (`mcp/server/lowlevel/server.py:713-717`) and needs the transport to push an `Exception` object into the read stream.

I searched the `mcp` package only, and I do not have `fastmcp==3.4.7` itself to check, so treat that as what I found rather than a closed list. Hence the question: which mcp version was the capture behind the 2026-08-25 validation taken against? It bears on two things.

`test_demotes_real_task_group_exception_group` documents its shape as the one actually logged in production and names `_send_response` as the source — which is exactly the call whose two entry points are guarded above.

And it decides the leaf set. That guard's own comment reads "Broken if the peer end closed first (`streamable_http terminate()`)", and `anyio==4.10.0` gives the same answer independently: both transport streams are unbuffered (`streamable_http.py:998-999`), so a send parks in `waiting_senders` whenever the message router is not currently parked in receive; `terminate()` closes `_write_stream_reader` (`:805`) before `_write_stream` (`:807`); `MemoryObjectReceiveStream.close()` wakes every parked sender; and the woken `send()` decides on `waiting_senders` membership alone, so it raises `BrokenResourceError` even though the send stream is closed by then. `_is_only_closed_resource_errors` (`src/ha_mcp/log_filters.py:102`) does not match that, since the two are sibling `Exception` subclasses. Is there a reason to leave it out of the leaf set? This repo already pairs them for streamable-HTTP memory-stream teardown in `tests/src/haos_runtime.py:1788-1789`.

**[Concern 2]: the silent older-server branch has no test.**

`_install_log_filters_if_available` deliberately splits `ModuleNotFoundError` for exactly `ha_mcp.log_filters` (silent) from everything else (WARNING), and the docstring is explicit that only the first is the older-server case. `test_serve_tolerates_missing_log_filters_module` stubs a module that exists but lacks the attribute, so it lands in the WARNING branch, and it asserts only that serving continues; no test in the suite constructs a `ModuleNotFoundError` for that module. Dropping the `err.name` condition therefore leaves the suite green while every older install gains a startup WARNING, which is the class of noise this PR exists to remove. The discriminating assertion is about output rather than survival: no warning on the older-server path, one on a genuinely missing dependency. Your `_stub_ha_mcp_surface` comment already works out why the `sys.modules` entry cannot simply be dropped, so the harder half is solved.

Neither is a regression; both fall back to today's behaviour.

**Comment by swissmo** [comment]
> The recursive leaf test is the right shape here, and the mixed-group case staying at ERROR is what makes it safe to ship: a demotion rule that swallowed a real bug travelling alongside the disconnect would be worse than the noise it removes. Pulling all three filters into a side-effect-free module rather than only the new one is also the right call, since the embedded server was missing all of them.
> 
> ...

Thank you for your review, much appreciated.

Both fixes pushed (df96851b):

**1. `BrokenResourceError` recognized** — `_is_only_closed_resource_errors` renamed to `_is_only_disconnect_teardown_errors`, now matches both `ClosedResourceError` and `BrokenResourceError` as an equivalent pair (matching the precedent already in `tests/src/haos_runtime.py`). 8 new/updated test cases covering bare, grouped, nested, and mixed-both-benign-types shapes.
**2. Silent-branch coverage closed** — new `TestInstallLogFiltersIfAvailable` calls `_install_log_filters_if_available()` directly with `caplog`, covering all three real branches: genuinely-missing-module (silent, using the same `__path__ = []` sys.modules technique the existing harness already established), attribute-missing (warns), and a different missing dependency (warns, names it).

**Review by Patch76 — APPROVED** [review]
Both fixes verified at b72550d3.

`_is_only_disconnect_teardown_errors` now pairs the two errors in `_DISCONNECT_TEARDOWN_ERRORS` (`src/ha_mcp/log_filters.py:88`), and the docstring records why they are one situation rather than two — the part that stops the next reader from tidying one of them away. Widening a demotion rule's leaf set is where the safety property tends to disappear quietly, so that is what I checked hardest: it did not. `test_leaves_mixed_exception_group_at_error` still asserts `record.exc_info is original_exc_info`, and the recursion still requires `all(...)` over a non-empty group.

The silent-branch coverage asserts on output rather than survival, which is what makes it load-bearing: `caplog.records == []` in `test_genuinely_missing_module_is_silent`, a warning for attribute-missing, a warning naming the dependency for the third. Dropping the `err.name` condition would fail that first assertion.

The switch in 63cf083d from an empty `__path__` to patching `builtins.__import__` is worth more than it looks. With the editable install's meta-path finder resolving `ha_mcp.*` by name, the earlier version could have raised nothing at all and passed vacuously — a green test for a branch never entered.

One non-blocking note, and it turns out to be about SDK drift rather than about your capture. The docstring of `test_demotes_real_task_group_exception_group` (`tests/src/unit/test_session_disconnect_log_filter.py:126`) attributes the shape to `_send_response`. That fits an older SDK exactly: at `mcp==1.24.0`, `Server._handle_request` awaits `message.respond(response)` with no `try` around it (`mcp/server/lowlevel/server.py:770`), so a disconnect error from that send escapes into `run()`'s task group and arrives at `streamable_http_manager`'s `logger.exception` as a nested `ExceptionGroup` — the shape the filter targets. At the `mcp==1.28.1` that `uv.lock` pins, the same send is caught against exactly `ClosedResourceError` and `BrokenResourceError` and dropped at DEBUG (`mcp/server/lowlevel/server.py:806-815`), so that particular route is closed; `_send_response`'s other caller, `RequestResponder.cancel()` (`mcp/shared/session.py:148`), has its only call site inside `_receive_loop`, where a raise is caught and logged as a validation warning. Nothing to change in the filter — it gates on the record's logger name, the presence of `exc_info`, and the leaf types, never on the SDK's wording, so it stays correct on both. Worth one clause in that docstring, though, since a reader on the newer SDK will try to follow the attribution and find the guard instead.

Approving on that basis — the demotion is correctly scoped, the mixed-group escape hatch is pinned by a test, and all three launchers now install the same filter set.
